src/mlpack/core/tree/cellbound_impl.hpp

```diff
@@ -0,0 +1,844 @@
+      // of hyperrectangles that we have added previously.
+      tmpLoAddress[row] ^= (AddressElemType) 1 << bit;
+
+      addr::AddressToPoint(loCorner, tmpLoAddress);
+      addr::AddressToPoint(hiCorner, tmpHiAddress);
+
+      AddBound(loCorner, hiCorner);
+    }
+
+    // The lower bound should correspond to this step.
+    tmpLoAddress[row] &= ~((AddressElemType) 1 << bit);
+  }
+}
+
+template<typename MetricType, typename ElemType>
+void CellBound<MetricType, ElemType>::UpdateAddressBounds()
+{
+  numBounds = 0;
+
+  // Calculate the number of equal leading bits of the lower address and
+  // the high address.
+  size_t row = 0;
+  for ( ; row < hiAddress.n_elem; row++)
+    if (loAddress[row] != hiAddress[row])
+      break;
+
+  // If the high address is equal to the lower address.
+  if (row == hiAddress.n_elem)
+  {
+    for (size_t i = 0; i < dim; i++)
+    {
+      loBound(i, 0) = bounds[i].Lo();
+      hiBound(i, 0) = bounds[i].Hi();
+    }
+    numBounds = 1;
+
+    return;
+  }
+
+  size_t bit = 0;
+  for ( ; bit < order; bit++)
+    if ((loAddress[row] & ((AddressElemType) 1 << (order - 1 - bit))) !=
+        (hiAddress[row] & ((AddressElemType) 1 << (order - 1 - bit))))
+      break;
+
+  if ((row == hiAddress.n_elem - 1) && (bit == order - 1))
+  {
+    // If the addresses differ in the last bit.
+    for (size_t i = 0; i < dim; i++)
+    {
+      loBound(i, 0) = bounds[i].Lo();
+      hiBound(i, 0) = bounds[i].Hi();
+    }
+
+    numBounds = 1;
+
+    return;
+  }
+
+  size_t numEqualBits = row * order + bit;
+  InitHighBound(numEqualBits);
+  InitLowerBound(numEqualBits);
+
+  assert(numBounds <= maxNumBounds);
+
+  if (numBounds == 0)
+  {
+    // I think this should never happen.
+    for (size_t i = 0; i < dim; i++)
+    {
+      loBound(i, 0) = bounds[i].Lo();
+      hiBound(i, 0) = bounds[i].Hi();
+    }
+
+    numBounds = 1;
+  }
+}
+
+/**
+ * Calculates minimum bound-to-point squared distance.
+ */
+template<typename MetricType, typename ElemType>
+template<typename VecType>
+inline ElemType CellBound<MetricType, ElemType>::MinDistance(
+    const VecType& point,
+    typename boost::enable_if<IsVector<VecType>>* /* junk */) const
+{
+  Log::Assert(point.n_elem == dim);
+
+  ElemType minSum = std::numeric_limits<ElemType>::max();
+
+  ElemType lower, higher;
+
+  for (size_t i = 0; i < numBounds; i++)
+  {
+    ElemType sum = 0;
+
+    for (size_t d = 0; d < dim; d++)
+    {
+      lower = loBound(d, i) - point[d];
+      higher = point[d] - hiBound(d, i);
+
+      // Since only one of 'lower' or 'higher' is negative, if we add each's
+      // absolute value to itself and then sum those two, our result is the
+      // nonnegative half of the equation times two; then we raise to power Power.
+      sum += pow((lower + fabs(lower)) + (higher + fabs(higher)),
+          (ElemType) MetricType::Power);
+    }
+
+    if (sum < minSum)
+      minSum = sum;
+  }
+
```

Comment:
So you have two different bounding structures, if I understand right: the minimum bounding hyperrectangle, and also the series of bounding rectangles defined by the minimum and maximum addresses.  In some cases that minimum bounding hyperrectangle will be much tighter than the series of bounding rectangles, so I wonder: you might get better performance if this function took the max of the minimum distances to the series of bounding rectangles and to the minimum bounding rectangle.  (The same applies to the other bounding functions too.)